server/sonar-server/src/main/java/org/sonar/server/es/queue/RecoveryIndexer.java

```diff
@@ -0,0 +1,162 @@
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.lang.math.RandomUtils;
+import org.sonar.api.Startable;
+import org.sonar.api.config.Settings;
+import org.sonar.api.utils.System2;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.api.utils.log.Profiler;
+import org.sonar.core.util.stream.MoreCollectors;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.es.EsQueueDto;
+import org.sonar.server.user.index.UserIndexer;
+
+import static java.lang.String.format;
+
+public class RecoveryIndexer implements Startable {
+
+  private static final Logger LOG = Loggers.get(RecoveryIndexer.class);
+  private static final String PROPERTY_INITIAL_DELAY = "sonar.search.recovery.initialDelayInSeconds";
+  private static final String PROPERTY_DELAY = "sonar.search.recovery.delayInSeconds";
+  private static final String PROPERTY_MIN_AGE = "sonar.search.recovery.minAgeInSeconds";
+  private static final String PROPERTY_LOOP_LIMIT = "sonar.search.recovery.loopLimit";
+  private static final long DEFAULT_DELAY_IN_SECONDS = 5L * 60;
+  private static final long DEFAULT_MIN_AGE_IN_SECONDS = 5L * 60;
+  private static final int DEFAULT_LOOP_LIMIT = 10_000;
+  private static final double FAILURE_CIRCUIT_BREAKER_IN_PERCENT = 0.3;
+
+  // Not static for testing purposes
+  private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1,
+    new ThreadFactoryBuilder()
+      .setPriority(Thread.MIN_PRIORITY)
+      .setNameFormat("RecoveryIndexer-%d")
+      .build());
+  private final System2 system2;
+  private final Settings settings;
+  private final DbClient dbClient;
+  private final UserIndexer userIndexer;
+
+  // initialized at constructor
+  private long minAgeInSeconds;
+  private final long loopLimit;
+
+  public RecoveryIndexer(System2 system2, Settings settings, DbClient dbClient, UserIndexer userIndexer) {
+    this.system2 = system2;
+    this.settings = settings;
+    this.dbClient = dbClient;
+    this.userIndexer = userIndexer;
+    this.minAgeInSeconds = getSetting(PROPERTY_MIN_AGE, DEFAULT_MIN_AGE_IN_SECONDS);
+    this.loopLimit = getSetting(PROPERTY_LOOP_LIMIT, DEFAULT_LOOP_LIMIT);
+  }
+
+  @Override
+  public void start() {
+    long delayInSeconds = getSetting(PROPERTY_DELAY, DEFAULT_DELAY_IN_SECONDS);
+
+    // in the cluster mode, avoid (but not prevent!) simultaneous executions of recovery
+    // indexers so that a document is not handled multiple times.
+    long initialDelay = getSetting(PROPERTY_INITIAL_DELAY, RandomUtils.nextInt(1 + (int) (delayInSeconds / 2)));
+
+    executorService.scheduleAtFixedRate(
+      this::index,
+      initialDelay,
+      delayInSeconds,
+      TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void stop() {
+    try {
+      executorService.shutdown();
+      executorService.awaitTermination(5, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either re-interrupt this method or rethrow the "InterruptedException". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2142)